File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterModelAssociation.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore.types.model

import com.amazonaws.amplify.cast
import com.amplifyframework.core.model.ModelAssociation

data class FlutterModelAssociation(val map: Map<String, Any>) {

    private val associationType: String = map["associationType"] as String
    private val targetNames = when (val targetNames = map["targetNames"]) {
        is List<*> -> targetNames.cast()
        else -> when (val targetName = map["targetName"]) {
            is String -> listOf(targetName)
            else -> null
        }
    }
    private val associatedName: String? = map["associatedName"] as String?
    private val associatedType: String? = map["associatedType"] as String?

    fun convertToNativeModelAssociation(): ModelAssociation {
        return ModelAssociation.builder()
            .name(associationType)
            .targetNames(targetNames?.toTypedArray())
            .associatedName(associatedName)
            .associatedType(associatedType)
            .build()
    }
}

File: packages/amplify_datastore/ios/Classes/types/model/FlutterModelAssociation.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Amplify
import amplify_flutter_ios

public struct FlutterModelAssociation {
    private let associationType: String
    private var targetNames = [String]()
    private let associatedName: String?

    init(serializedData: [String: Any]) throws {
        guard let associationType = serializedData["associationType"] as? String
        else {
            throw ModelSchemaError.parse(
                className: "FlutterModelAssociation",
                fieldName: "associationType",
                desiredType: "String")
        }
        self.associationType = associationType

        // With custom primary key we are expecting model to provide targetNames instead of targetName
        // For belongsTo model association, if this information is missing, mostl likely
        // Flutter sent wrongly serialized data
        // Also, keeping targetName for backwards compatibility
        if let inputTargetName = serializedData["targetName"] as? String {
            targetNames.append(inputTargetName)
        } else if let inputTargetNames = serializedData["targetNames"] as? [String] {
            targetNames += inputTargetNames
        } else if associationType == "BelongsTo" {
            throw DataStoreError.decodingError(
                "Invalid ModelAssociation data: targetNames is not presented in the serialized ModelAssociation data",
                ErrorMessages.missingAttribute)
        }
        self.associatedName = serializedData["associatedName"] as? String
    }

    public func convertToNativeModelAssociation() -> ModelAssociation{
        switch associationType {
            case "HasMany":
                return .hasMany(associatedFieldName: associatedName)
            case "HasOne":
                return .hasOne(associatedFieldName: associatedName, targetNames: targetNames)
            case "BelongsTo":
                return .belongsTo(associatedFieldName: associatedName, targetNames: targetNames)
            default:
                preconditionFailure("Could not create a ModelAssociation from \(associationType)")
        }
    }
}
